lib/tensorflow/graph.rb

```diff
@@ -66,18 +67,27 @@ def define_op(opName, name , input, device, attrs)
     op = self.available_ops[opName.downcase]
     raise ("Operation does not exist.") if !op
     opName = op.name   # This ensures that case-sensitivity does not become an issue
+    input ||= []
     raise ("Invalid number of inputs.") if op.input_arg.length != input.length
     inputs = []
-    input.each do |node|
-      inputs.push(node.definition.name)
+    (0..input.length - 1).each do |i|
+      begin
+        if op.input_arg[i].is_ref and input[i].ref
```

Comment:
Never use `and` in this context, only `&&`. `and` has unexpected precedence and is not a logical operator, it is code flow operators (like `do_something and return`).